Change of CovertCSVtoRecords for multi region CSV formats

Hijlko opened the pull request:
I donwloaded the package for converting CSV files to records. I found it very usefull but had to change the apex code to be able to import CSV files of the European region. Standard US csv uses a comma as a field separator while most European csv's use a semicolon as a field seperator. There are more diferences between the regions.
Date: YYYY-MM-DD for US and SalesForce standard but DD-MM-YYYY for European standard
Thousand/decimal separator: In US the point is used as a decimal separator an a comma as a thousands separator.
In Europe a comma is used as a decimal separator and a point as thousands separator.

That's why  I made some changes to the code/componet so you're able to set the 
- field separator
- Thousand seperator
- Decimal Separator and
- Currency symbol

The code can now import muli region CSV files
Numbers/currency like 1,2345.56 (US) and 1.234,56 (EU) are converted to 1234.56
Date like YYYY-MM-DD and DD-MM-YYYY (and other separators like YYYY/MM/DD and DD/MM/YYYY) are always converted to YYYY-MM-DD


I'm very new to APEX coding so maybe it can be coded in a better way but it works for me. Maybe you can check it out.
[ConvertAllCSVToRecords.txt](https://github.com/alexed1/LightningFlowComponents/files/6093756/ConvertAllCSVToRecords.txt)

alexed1 commented:
@Hijlko thanks for this great improvement!

The first step is we need to adjust your pull request. Note that it says that it is changing 219 files:
![image](https://user-images.githubusercontent.com/3140883/110217313-25c07300-7e68-11eb-98af-8bfebed8563c.png)

We don't want that. You're really only changing 1 or 2 files. It looks like you're trying to merge into the 3.0.9 branch, which is a branch someone created for a particular reason. You want to merge into the 'master' branch. 

There's no damage done if you create a pull request like this one and it's not right. Just close or abandon it and create a new one. You'll want to read up a little on pull requests and branches. If you want to provide an email address, I'll invite you to a slack workspace used by flow developers and we can help you get going.

Hijlko commented:
Hi, It was really not clear for me how to create a new pull request. The new request button was grayed out. So I still don't know how.
Kind regards.
Hijlko
________________________________
From: Alex Edelstein <notifications@github.com>
Sent: Saturday, March 6, 2021 7:42:55 PM
To: alexed1/LightningFlowComponents <LightningFlowComponents@noreply.github.com>
Cc: Hijlko <hijlko.roosjen@outlook.com>; Mention <mention@noreply.github.com>
Subject: Re: [alexed1/LightningFlowComponents] Change of CovertCSVtoRecords for multi region CSV formats (#626)


@Hijlko<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FHijlko&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764860640%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=VuPn9hfcsErX%2FR8Cl1tRknDSFyvpqEgpSNTylGOCyOs%3D&reserved=0> thanks for this great improvement!

The first step is we need to adjust your pull request. Note that it says that it is changing 219 files:
[image]<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F3140883%2F110217313-25c07300-7e68-11eb-98af-8bfebed8563c.png&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764860640%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=HONJi52FwOwf9Ojo%2FpQMad20HzYRSkg%2B8WwFkrBjEUk%3D&reserved=0>

We don't want that. You're really only changing 1 or 2 files. It looks like you're trying to merge into the 3.0.9 branch, which is a branch someone created for a particular reason. You want to merge into the 'master' branch.

There's no damage done if you create a pull request like this one and it's not right. Just close or abandon it and create a new one. You'll want to read up a little on pull requests and branches. If you want to provide an email address, I'll invite you to a slack workspace used by flow developers and we can help you get going.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Falexed1%2FLightningFlowComponents%2Fpull%2F626%23issuecomment-792015914&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764870642%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=pAqKwfs3Hbd003tvjTp0e2Pcjk4tVcn2ql7JobBzqrk%3D&reserved=0>, or unsubscribe<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FATDLIKYAYMKEX3BMM2TMAL3TCJZS7ANCNFSM4YV7SJ6A&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764870642%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=C7bCbnbiVyvNjZwGDrZ3Lzw%2Ftc2yYxfmhO1twb%2FTDIU%3D&reserved=0>.

ericrsmith35 commented:
Hijlko,

How many files required changes for your update?  I can try and help you
get a valid Pull Request created if you like.

Eric Smith

On Sat, Mar 6, 2021 at 2:47 PM Hijlko <notifications@github.com> wrote:

> Hi, It was really not clear for me how to create a new pull request. The
> new request button was grayed out. So I still don't know how.
> Kind regards.
> Hijlko
> ________________________________
> From: Alex Edelstein <notifications@github.com>
> Sent: Saturday, March 6, 2021 7:42:55 PM
> To: alexed1/LightningFlowComponents <
> LightningFlowComponents@noreply.github.com>
> Cc: Hijlko <hijlko.roosjen@outlook.com>; Mention <
> mention@noreply.github.com>
> Subject: Re: [alexed1/LightningFlowComponents] Change of
> CovertCSVtoRecords for multi region CSV formats (#626)
>
>
> @Hijlko<
> https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FHijlko&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764860640%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=VuPn9hfcsErX%2FR8Cl1tRknDSFyvpqEgpSNTylGOCyOs%3D&reserved=0>
> thanks for this great improvement!
>
> The first step is we need to adjust your pull request. Note that it says
> that it is changing 219 files:
> [image]<
> https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F3140883%2F110217313-25c07300-7e68-11eb-98af-8bfebed8563c.png&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764860640%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=HONJi52FwOwf9Ojo%2FpQMad20HzYRSkg%2B8WwFkrBjEUk%3D&reserved=0
> >
>
> We don't want that. You're really only changing 1 or 2 files. It looks
> like you're trying to merge into the 3.0.9 branch, which is a branch
> someone created for a particular reason. You want to merge into the
> 'master' branch.
>
> There's no damage done if you create a pull request like this one and it's
> not right. Just close or abandon it and create a new one. You'll want to
> read up a little on pull requests and branches. If you want to provide an
> email address, I'll invite you to a slack workspace used by flow developers
> and we can help you get going.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub<
> https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Falexed1%2FLightningFlowComponents%2Fpull%2F626%23issuecomment-792015914&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764870642%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=pAqKwfs3Hbd003tvjTp0e2Pcjk4tVcn2ql7JobBzqrk%3D&reserved=0>,
> or unsubscribe<
> https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FATDLIKYAYMKEX3BMM2TMAL3TCJZS7ANCNFSM4YV7SJ6A&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764870642%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=C7bCbnbiVyvNjZwGDrZ3Lzw%2Ftc2yYxfmhO1twb%2FTDIU%3D&reserved=0
> >.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/alexed1/LightningFlowComponents/pull/626#issuecomment-792040353>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF4UHX4NNUN4RLXJMBTCCWDTCKBE3ANCNFSM4YV7SJ6A>
> .
>

Hijlko commented:
Hi Eric,

Thanks.

It was just 1. I saw the CovertCSVtoRecords and thought it was from Narender SIngh. My code is a modification to allow US and EU style csv files to be imported.
There still maybe an error in the uploaded code becaude it gave me an error about assigning a null value to a variable. I Changed my version to a single CSV file upload today because I only need to convert 1 csv file at a time. After that, the error was gone. Now I still want to adjust the test class for it. There is 1 error in that:
System.AssertException: Assertion Failed
Stack Trace: Class.ConvertAllCSVToRecordsTest.testconvertMethodExceptions
The changes I made to the apex class I changed today and the testclass is attached to this email.
Kind regards,\
Hijlko Roosjen
The Netherlands
________________________________
Van: Eric Smith <notifications@github.com>
Verzonden: zondag 7 maart 2021 21:44
Aan: alexed1/LightningFlowComponents <LightningFlowComponents@noreply.github.com>
CC: Hijlko <hijlko.roosjen@outlook.com>; Mention <mention@noreply.github.com>
Onderwerp: Re: [alexed1/LightningFlowComponents] Change of CovertCSVtoRecords for multi region CSV formats (#626)


Hijlko,

How many files required changes for your update? I can try and help you
get a valid Pull Request created if you like.

Eric Smith

On Sat, Mar 6, 2021 at 2:47 PM Hijlko <notifications@github.com> wrote:

> Hi, It was really not clear for me how to create a new pull request. The
> new request button was grayed out. So I still don't know how.
> Kind regards.
> Hijlko
> ________________________________
> From: Alex Edelstein <notifications@github.com>
> Sent: Saturday, March 6, 2021 7:42:55 PM
> To: alexed1/LightningFlowComponents <
> LightningFlowComponents@noreply.github.com>
> Cc: Hijlko <hijlko.roosjen@outlook.com>; Mention <
> mention@noreply.github.com>
> Subject: Re: [alexed1/LightningFlowComponents] Change of
> CovertCSVtoRecords for multi region CSV formats (#626)
>
>
> @Hijlko<
> https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FHijlko&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764860640%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=VuPn9hfcsErX%2FR8Cl1tRknDSFyvpqEgpSNTylGOCyOs%3D&reserved=0>
> thanks for this great improvement!
>
> The first step is we need to adjust your pull request. Note that it says
> that it is changing 219 files:
> [image]<
> https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F3140883%2F110217313-25c07300-7e68-11eb-98af-8bfebed8563c.png&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764860640%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=HONJi52FwOwf9Ojo%2FpQMad20HzYRSkg%2B8WwFkrBjEUk%3D&reserved=0
> >
>
> We don't want that. You're really only changing 1 or 2 files. It looks
> like you're trying to merge into the 3.0.9 branch, which is a branch
> someone created for a particular reason. You want to merge into the
> 'master' branch.
>
> There's no damage done if you create a pull request like this one and it's
> not right. Just close or abandon it and create a new one. You'll want to
> read up a little on pull requests and branches. If you want to provide an
> email address, I'll invite you to a slack workspace used by flow developers
> and we can help you get going.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub<
> https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Falexed1%2FLightningFlowComponents%2Fpull%2F626%23issuecomment-792015914&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764870642%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=pAqKwfs3Hbd003tvjTp0e2Pcjk4tVcn2ql7JobBzqrk%3D&reserved=0>,
> or unsubscribe<
> https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FATDLIKYAYMKEX3BMM2TMAL3TCJZS7ANCNFSM4YV7SJ6A&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764870642%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=C7bCbnbiVyvNjZwGDrZ3Lzw%2Ftc2yYxfmhO1twb%2FTDIU%3D&reserved=0
> >.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/alexed1/LightningFlowComponents/pull/626#issuecomment-792040353>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF4UHX4NNUN4RLXJMBTCCWDTCKBE3ANCNFSM4YV7SJ6A>
> .
>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Falexed1%2FLightningFlowComponents%2Fpull%2F626%23issuecomment-792349444&data=04%7C01%7C%7C1b13c6f4dea44162b31008d8e1a9de76%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637507466992815466%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=GSG7yY4vdfYUlSl8J%2Fuw82KYccleU9QNeiKd8fL34E4%3D&reserved=0>, or unsubscribe<https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FATDLIK3ATVQIWVW5ABDCJMDTCPQURANCNFSM4YV7SJ6A&data=04%7C01%7C%7C1b13c6f4dea44162b31008d8e1a9de76%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637507466992825464%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=vY1TeEpx6xydAxQ%2BneckKtMI21vnZ%2B7oaMNy5n3Xq3Y%3D&reserved=0>.

public with sharing class ConvertAllCSVToRecords {

    @InvocableMethod(label='Convert CSV File to Records from single ContentDocumentId')
    public static Response[] convert (Request[] inputs) {
        Response[] responseList = new Response[]{};
        
/*        for (Request curInput : inputs) {
    
            if (curInput.contentDocumentId == null && curInput.contentDocumentIdList == null) {
                throw new FlowCustomException('You need to provide either a single ContentDocument Id OR a list of ContentDocumentIds as inputs to this Flow action.');
            }
            else {
                // Easiest to just combine the inputs for processing
                if (curInput.contentDocumentId != null)
                    curInput.contentDocumentIdList.add(curInput.contentDocumentId);
            }
            for (String curContentDocumentId : curInput.contentDocumentIdList){
*/        for (Request curInput : inputs) {
            if (curInput.contentDocumentId == null) {
                throw new FlowCustomException('You need to provide a single ContentDocument Id.');
            }
            List<String> contentDocumentIdList = new List<String> {curInput.contentDocumentId};
            for (String curContentDocumentId : contentDocumentIdList){
                SObjectType sObjType = ((SObject)Type.forName(curInput.objectApiName).newInstance()).getSObjectType();         
                map<String,SObjectField> fieldMap = sObjType.getDescribe().fields.getMap();

                Blob csvBlobContent = getCsvContent(curContentDocumentId);
                String csvStringContent = blobToString(csvBlobContent, 'ISO-8859-1');
                csvStringContent = csvStringContent.replaceAll('(\r\n|\r)','\n');

                String[] csvRows = csvStringContent.split('\n');
                String[] fields = csvRows[0].split(curInput.FSep); //Gives all the API names of the fields from the 1st CSV row 
                csvRows.remove(0);
                map<String,Schema.DisplayType> fieldToFieldTypeMap = new map<String,Schema.DisplayType>();
                map<String,String> fieldName = new map<String,String>();
                for (String field : fields) {
                    String fieldApiName = field;
                    if(fieldMap.get(fieldApiName) == null) {
                        fieldApiName = fieldApiName.replaceAll(' ', '_').replaceAll('[^a-zA-Z0-9_]', '');   // Convert to alphanumeric and underscore
                        fieldApiName = fieldApiName.replaceAll('__', '_')+'__c';                            // Remove extra _s and add __c to end  
                    }                       
                    fieldToFieldTypeMap.put(field, getFieldType(fieldMap, fieldApiName));
                    fieldName.put(field, fieldApiName); 
                }

                Response output = new Response();
                output.convertedCSVRows = new SObject[]{};
                String tempRepl = '~`~';
                for(String row : csvRows) {
                    SObject obj = sObjType.newSObject();
                    row = row.replace(curInput.Fsep + '"',curInput.Fsep).replace('"' + curInput.Fsep,curInput.Fsep);
                    String[] fieldInputValues = row.split(curInput.FSep);
                    for (Integer i = 0; i < fields.size(); i++) {
                        String fieldValue = fieldInputValues[i].replace(curInput.Fsep + '"',curInput.Fsep).replace('"' + curInput.Fsep,curInput.Fsep);
                        String Tsep = curInput.Tsep;
                        String Dsep = curInput.Dsep;
                        String Csymb = curInput.Csymb;
                        Schema.DisplayType fieldType = fieldToFieldTypeMap.get(fields[i]);
                        try {
                            obj.put( fieldName.get(fields[i]), getConvertedFieldValue(fieldType, fieldValue, Tsep, Dsep, Csymb) );
                        } catch(exception e) {
                            obj.put( fieldName.get(fields[i]), null );
                        }
                    }
                    output.convertedCSVRows.add(obj);
                }
                responseList.add(output);
            }            
            

        }
        return responseList;
    }

    static Blob getCsvContent( String contentDocumentId) {        
        ContentVersion csvFile = [SELECT 
                                    VersionData 
                                    FROM ContentVersion 
                                    WHERE ContentDocumentId = :contentDocumentId 
                                    AND IsLatest = true
                                ];
        System.debug('contentDocumentId is: ' + contentDocumentId);
        System.debug('csvFile is: ' + csvFile);
        return csvFile.VersionData;
    }

    /* Converting the CSV file input(BLOB format) to a string */
    static String blobToString(Blob input, String inCharset){
        String hex = EncodingUtil.convertToHex(input);
        System.assertEquals(0, hex.length() & 1);
        final Integer bytesCount = hex.length() >> 1;
        String[] bytes = new String[bytesCount];
        for(Integer i = 0; i < bytesCount; ++i)
            bytes[i] =  hex.mid(i << 1, 2);
        return EncodingUtil.urlDecode('%' + String.join(bytes, '%'), inCharset);
    } 

    static Object getConvertedFieldValue(Schema.DisplayType fieldType, String fieldValue, String Tsep, String Dsep, String Csymb){
        Object convertedFieldValue;

        switch on fieldType {//Setting Field value based on the field data type
            when BOOLEAN {
                convertedfieldValue = String.isEmpty(fieldValue) ? false : Boolean.valueOf(fieldValue);
            }
            when DOUBLE {
                String fValue = fieldValue.replaceAll('^0-9,.','');
                if( fValue.indexOf(Tsep) < 0 && fValue.indexOf(Dsep) < 0 ) {/* both WITHOUT Thousands separator and WITHOUT decimal separator */
                    /* No conversion needed */
                    convertedfieldValue = String.isEmpty(fieldValue) ? null : Decimal.valueOf(fValue);
                    }
                if( fValue.indexOf(Tsep) < 0 && fValue.indexOf(Dsep) > -1) {/*example: change value WITHOUT thousands separator and WITH decimals like 1.234 (US) or 1,234 (EU) to 1.234 (US)*/
                    /* Convert comma as decimal separator to point */
                    convertedfieldValue = String.isEmpty(fieldValue) ? null : Decimal.valueOf(fValue.replace(',','.'));
                    }
                if( fValue.indexOf(Tsep) > -1 && fValue.indexOf(Dsep) < 0) {/*example: change value WITH thousands separator and WITHOUT decimals like 1,234 (US) or 1.234 (EU) to 1234 */
                    /* Remove comma and point as Thousands separator (no decimals present) */
                    convertedfieldValue = String.isEmpty(fieldValue) ? null : Decimal.valueOf(fValue.replace(',','').replace('.',''));
                    }
                if( fValue.indexOf(Tsep) > -1 && fValue.indexOf(Dsep) > -1 && Tsep == ',') {/*example: change value WITH thousand separator and WITH decimals like 1,234.56 to 1234.56 (US) */
                    /* Remove comma Thousands separator, leave decimal separator unchanged) */
                    convertedfieldValue = String.isEmpty(fieldValue) ? null : Decimal.valueOf(fValue.replace(',',''));
                    }
                if( fValue.indexOf(Tsep) > -1 && fValue.indexOf(Dsep) > -1 && Tsep == '.') {/*example: change value WITH thousand separator and WITH decimals like 1.234,56 to 1234,56 (EU)*/
                    /* Remove point as Thousands separator, change comma to point as decimal separator */
                    convertedfieldValue = String.isEmpty(fieldValue) ? null : Decimal.valueOf(fValue = fValue.replace('.','').replace(',','.'));
                    }
            }
            when CURRENCY {
                String fValue = fieldValue.replaceAll('^0-9,.','').replace(Csymb,'').replace(' ','');
                if( fValue.indexOf(Tsep) < 0 && fValue.indexOf(Dsep) < 0 ) {/* both WITHOUT Thousands separator and WITHOUT decimal separator */
                    /* No conversion needed */
                    convertedfieldValue = String.isEmpty(fieldValue) ? null : Decimal.valueOf(fValue);
                    }
                if( fValue.indexOf(Tsep) < 0 && fValue.indexOf(Dsep) > -1) {/*example: change value WITHOUT thousands separator and WITH decimals like 1.234 (US) or 1,234 (EU) to 1.234 (US)*/
                    /* Convert comma as decimal separator to point */
                    convertedfieldValue = String.isEmpty(fieldValue) ? null : Decimal.valueOf(fValue.replace(',','.'));
                    }
                if( fValue.indexOf(Tsep) > -1 && fValue.indexOf(Dsep) < 0) {/*example: change value WITH thousands separator and WITHOUT decimals like 1,234 (US) or 1.234 (EU) to 1234 */
                    /* Remove comma and point as Thousands separator (no decimals present) */
                    convertedfieldValue = String.isEmpty(fieldValue) ? null : Decimal.valueOf(fValue.replace(',','').replace('.',''));
                    }
                if( fValue.indexOf(Tsep) > -1 && fValue.indexOf(Dsep) > -1 && Tsep == ',') {/*example: change value WITH thousand separator and WITH decimals like 1,234.56 to 1234.56 (US) */
                    /* Remove comma Thousands separator, leave decimal separator unchanged) */
                    convertedfieldValue = String.isEmpty(fieldValue) ? null : Decimal.valueOf(fValue.replace(',',''));
                    }
                if( fValue.indexOf(Tsep) > -1 && fValue.indexOf(Dsep) > -1 && Tsep == '.') {/*example: change value WITH thousand separator and WITH decimals like 1.234,56 to 1234,56 (EU)*/
                    /* Remove point as Thousands separator, change comma to point as decimal separator */
                    convertedfieldValue = String.isEmpty(fieldValue) ? null : Decimal.valueOf(fValue = fValue.replace('.','').replace(',','.'));
                    }
            }
            when INTEGER {
                /*Remove thousands separator ( , and . like in 1,234 and 1.234 ) */
                convertedfieldValue = String.isEmpty(fieldValue) ? null : Integer.valueOf(fieldValue.replaceAll('^0-9',''));
            }
            when DATE {
                /* replace / and : and . in fieldvalue by -  like in 2021.12.31 or 2021/12/31 or 2021:12:31 or 1.1.2021 or 31/12/2021 or 31:12:2021 */
                String fValue = fieldValue.replaceAll('^0-9','-');
                list<String> lstSplit = fValue.split('-');
                If (fValue.indexOf('-') > 2){
                   /* Format found like YYYY-MM-DD */
                   Integer Year = Integer.valueOf(lstSplit[0]);           
                   Integer Month = Integer.valueOf(lstSplit[1]);           
                   Integer Day = Integer.valueOf(lstSplit[2]);
                   convertedfieldValue = String.isEmpty(fieldValue) ? null : date.valueOf(Year + '-' + ('0' + Month).Right(2) + '-' + ('0' + Day).Right(2));
                   }
                else {
                   /* Format found like DD-MM-YYYY, convert to YYYY-MM-DD */
                   Integer Day = Integer.valueOf(lstSplit[0]);           
                   Integer Month = Integer.valueOf(lstSplit[1]);           
                   Integer Year = Integer.valueOf(lstSplit[2]);
                   convertedfieldValue = String.isEmpty(fieldValue) ? null : date.valueOf(Year + '-' + ('0' + Month).Right(2) + '-' + ('0' + Day).Right(2));
                   }                      
            }
            when else { 
                convertedfieldValue = fieldValue;
            }
        }

        return convertedFieldValue;
    }

    @TestVisible
    static Schema.DisplayType getFieldType(map<String, SObjectField> fieldMap, String fieldApiName) {
        if(fieldMap.get(fieldApiName) == null) {
            throw new FlowCustomException('Invalid Field Error: '+fieldApiName + '. Please double check the API name of the field.');
        }           
        return fieldMap.get(fieldApiName).getDescribe().getType();
    }
    
    public class FlowCustomException extends Exception{}

    public class Request{

        public Request() {
        }
        
        @InvocableVariable(label='Object API Name' required=true)
        public string objectApiName;

        @InvocableVariable(label='Separator 1: Fields' description='Choose comma (input: , ) or semicolon (input: ; )' required=true)
        public string FSep;

        @InvocableVariable(label='Separator 2: Decimals' description='Choose comma (input: , ) or semicolon (input: ; )' required=true)
        public string DSep;

        @InvocableVariable(label='Separator 3: Thousands' description='Choose comma (input: , ) or semicolon (input: ; )' required=true)
        public string TSep;
        
        @InvocableVariable(label='Symbol of Currency' description='Like $, €, £' required=true)
        public string Csymb;
        
        @InvocableVariable(label='CSV (single) File ID' required=true)
        public string contentDocumentId;

/*        @InvocableVariable(label='CSV List of File IDs')
        public List<String> contentDocumentIdList;
*/    }

    public class Response{
        @InvocableVariable(label='List(collection) of Records')
        public List<SObject> convertedCSVRows;
    }
}
@isTest
public with sharing class ConvertAllCSVToRecordsTest {
   
    
        @TestSetup
        static void createData(){
            
            ContentVersion cv = new ContentVersion(
                Title = 'Demo Accounts CSV',
                VersionData = createCSVBlob(),
                PathOnClient = 'DemoAccounts.csv'
            );
            insert cv;
        }
    
        @isTest
        static void testconvertMethodExceptions() {
            ConvertAllCSVToRecords.Request[] flowInputs = createFlowInputsList();
    
            /* 1. FlowInput list size more than 1 exception
             */
            try {
                ConvertAllCSVToRecords.Request input = new ConvertAllCSVToRecords.Request();
                flowInputs.add(input);
                ConvertAllCSVToRecords.Response[] flowOutputs = ConvertAllCSVToRecords.convert(flowInputs);
            } catch (Exception e) {
                System.assert(e.getMessage().contains('This action is only supported in Screen flows.'));
            }
        }
    
        @isTest
        static void testconvertMethod() {
    
            ConvertAllCSVToRecords.Request[] flowInputs = createFlowInputsList();
            Test.startTest();
            ConvertAllCSVToRecords.Response[] flowOutputs = ConvertAllCSVToRecords.convert(flowInputs);
            Test.stopTest();
            System.assert(flowOutputs[0].convertedCSVRows.size() == 3);
        }
    
        @isTest
        static void testGetFieldTypeMethodException() {
            try {
                ConvertAllCSVToRecords.getFieldType(Schema.SObjectType.Account.fields.getMap(),'TestField');
            } catch (Exception e) {
                System.assert(e.getMessage().contains('Invalid Field Error'));
            }
        }
    
        static Blob createCSVBlob() {
            String csvStringContent;
            Blob csvBlobContent;
            string[] fields = new String[]{
                'AccountNumber',
                'AnnualRevenue',
                'NumberOfEmployees',
                'Rating',
                'Name'
            };
            
            String csvHeader = String.join( fields, ',') + '\n';
            csvStringContent = csvHeader;
    
            Account[] accRecords = new Account[]{
                new Account(Name = 'CSV Acc1', AnnualRevenue = NULL, AccountNumber = '', NumberOfEmployees = NULL, Rating = 'Hot'),
                new Account(Name = 'CSV Acc2', AnnualRevenue = 1000, AccountNumber = '911', NumberOfEmployees = 0, Rating = 'Warm'),
                new Account(Name = 'CSV Acc3', AnnualRevenue = 911007, AccountNumber = '007',NumberOfEmployees = 10,Rating = 'Cold')
            };
            
            for(Integer i = 0; i < accRecords.size(); i++) {
                Account a = accRecords[i];
                for(Integer j = 0; j < fields.size(); j++) {
                    String field = fields[j];
                    csvStringContent = csvStringContent + (a.get(field) == NULL ? '' : a.get(field));
                    if( j < fields.size()-1) {
                        csvStringContent += + ',';
                    }
                }
                if( i < accRecords.size()-1) {
                    csvStringContent += '\n';
                }
            }
            return Blob.valueOf(csvStringContent);
        } 
    
        static ConvertAllCSVToRecords.Request[] createFlowInputsList(){
            String contentDocumentId;
            ContentVersion cv = [Select ContentDocumentId,FileType
                                 FROM ContentVersion
                                 Where Title = 'Demo Accounts CSV'
                                ];
            ConvertAllCSVToRecords.Request[] flowInputs = new ConvertAllCSVToRecords.Request[]{};
            ConvertAllCSVToRecords.Request input = new ConvertAllCSVToRecords.Request();
            input.objectApiName = 'Account';
            input.contentDocumentId = cv.ContentDocumentId;
            input.Fsep = ',';
            input.Tsep = ',';
            input.Dsep = '.';
            input.Csymb = '$';
            flowInputs.add(input);
    
            return flowInputs;
        }
    
    }

ericrsmith35 commented:
I've asked Narender to take a look at this.

On Mon, Mar 8, 2021 at 9:25 AM Hijlko <notifications@github.com> wrote:

> Hi Eric,
>
> Thanks.
>
> It was just 1. I saw the CovertCSVtoRecords and thought it was from
> Narender SIngh. My code is a modification to allow US and EU style csv
> files to be imported.
> There still maybe an error in the uploaded code becaude it gave me an
> error about assigning a null value to a variable. I Changed my version to a
> single CSV file upload today because I only need to convert 1 csv file at a
> time. After that, the error was gone. Now I still want to adjust the test
> class for it. There is 1 error in that:
> System.AssertException: Assertion Failed
> Stack Trace: Class.ConvertAllCSVToRecordsTest.testconvertMethodExceptions
> The changes I made to the apex class I changed today and the testclass is
> attached to this email.
> Kind regards,\
> Hijlko Roosjen
> The Netherlands
> ________________________________
> Van: Eric Smith <notifications@github.com>
> Verzonden: zondag 7 maart 2021 21:44
> Aan: alexed1/LightningFlowComponents <
> LightningFlowComponents@noreply.github.com>
> CC: Hijlko <hijlko.roosjen@outlook.com>; Mention <
> mention@noreply.github.com>
> Onderwerp: Re: [alexed1/LightningFlowComponents] Change of
> CovertCSVtoRecords for multi region CSV formats (#626)
>
>
> Hijlko,
>
> How many files required changes for your update? I can try and help you
> get a valid Pull Request created if you like.
>
> Eric Smith
>
> On Sat, Mar 6, 2021 at 2:47 PM Hijlko <notifications@github.com> wrote:
>
> > Hi, It was really not clear for me how to create a new pull request. The
> > new request button was grayed out. So I still don't know how.
> > Kind regards.
> > Hijlko
> > ________________________________
> > From: Alex Edelstein <notifications@github.com>
> > Sent: Saturday, March 6, 2021 7:42:55 PM
> > To: alexed1/LightningFlowComponents <
> > LightningFlowComponents@noreply.github.com>
> > Cc: Hijlko <hijlko.roosjen@outlook.com>; Mention <
> > mention@noreply.github.com>
> > Subject: Re: [alexed1/LightningFlowComponents] Change of
> > CovertCSVtoRecords for multi region CSV formats (#626)
> >
> >
> > @Hijlko<
> >
> https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FHijlko&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764860640%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=VuPn9hfcsErX%2FR8Cl1tRknDSFyvpqEgpSNTylGOCyOs%3D&reserved=0
> >
> > thanks for this great improvement!
> >
> > The first step is we need to adjust your pull request. Note that it says
> > that it is changing 219 files:
> > [image]<
> >
> https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F3140883%2F110217313-25c07300-7e68-11eb-98af-8bfebed8563c.png&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764860640%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=HONJi52FwOwf9Ojo%2FpQMad20HzYRSkg%2B8WwFkrBjEUk%3D&reserved=0
> > >
> >
> > We don't want that. You're really only changing 1 or 2 files. It looks
> > like you're trying to merge into the 3.0.9 branch, which is a branch
> > someone created for a particular reason. You want to merge into the
> > 'master' branch.
> >
> > There's no damage done if you create a pull request like this one and
> it's
> > not right. Just close or abandon it and create a new one. You'll want to
> > read up a little on pull requests and branches. If you want to provide an
> > email address, I'll invite you to a slack workspace used by flow
> developers
> > and we can help you get going.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub<
> >
> https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Falexed1%2FLightningFlowComponents%2Fpull%2F626%23issuecomment-792015914&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764870642%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=pAqKwfs3Hbd003tvjTp0e2Pcjk4tVcn2ql7JobBzqrk%3D&reserved=0
> >,
> > or unsubscribe<
> >
> https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FATDLIKYAYMKEX3BMM2TMAL3TCJZS7ANCNFSM4YV7SJ6A&data=04%7C01%7C%7Cdc8e037fcb574fd9da0208d8e0cfa828%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637506529764870642%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=C7bCbnbiVyvNjZwGDrZ3Lzw%2Ftc2yYxfmhO1twb%2FTDIU%3D&reserved=0
> > >.
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/alexed1/LightningFlowComponents/pull/626#issuecomment-792040353
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AF4UHX4NNUN4RLXJMBTCCWDTCKBE3ANCNFSM4YV7SJ6A
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub<
> https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Falexed1%2FLightningFlowComponents%2Fpull%2F626%23issuecomment-792349444&data=04%7C01%7C%7C1b13c6f4dea44162b31008d8e1a9de76%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637507466992815466%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=GSG7yY4vdfYUlSl8J%2Fuw82KYccleU9QNeiKd8fL34E4%3D&reserved=0>,
> or unsubscribe<
> https://emea01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FATDLIK3ATVQIWVW5ABDCJMDTCPQURANCNFSM4YV7SJ6A&data=04%7C01%7C%7C1b13c6f4dea44162b31008d8e1a9de76%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637507466992825464%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=vY1TeEpx6xydAxQ%2BneckKtMI21vnZ%2B7oaMNy5n3Xq3Y%3D&reserved=0
> >.
>
> public with sharing class ConvertAllCSVToRecords {
>
> @InvocableMethod(label='Convert CSV File to Records from single
> ContentDocumentId')
> public static Response[] convert (Request[] inputs) {
> Response[] responseList = new Response[]{};
>
> /* for (Request curInput : inputs) {
>
> if (curInput.contentDocumentId == null && curInput.contentDocumentIdList
> == null) {
> throw new FlowCustomException('You need to provide either a single
> ContentDocument Id OR a list of ContentDocumentIds as inputs to this Flow
> action.');
> }
> else {
> // Easiest to just combine the inputs for processing
> if (curInput.contentDocumentId != null)
> curInput.contentDocumentIdList.add(curInput.contentDocumentId);
> }
> for (String curContentDocumentId : curInput.contentDocumentIdList){
> */ for (Request curInput : inputs) {
> if (curInput.contentDocumentId == null) {
> throw new FlowCustomException('You need to provide a single
> ContentDocument Id.');
> }
> List<String> contentDocumentIdList = new List<String>
> {curInput.contentDocumentId};
> for (String curContentDocumentId : contentDocumentIdList){
> SObjectType sObjType =
> ((SObject)Type.forName(curInput.objectApiName).newInstance()).getSObjectType();
>
> map<String,SObjectField> fieldMap =
> sObjType.getDescribe().fields.getMap();
>
> Blob csvBlobContent = getCsvContent(curContentDocumentId);
> String csvStringContent = blobToString(csvBlobContent, 'ISO-8859-1');
> csvStringContent = csvStringContent.replaceAll('(\r\n|\r)','\n');
>
> String[] csvRows = csvStringContent.split('\n');
> String[] fields = csvRows[0].split(curInput.FSep); //Gives all the API
> names of the fields from the 1st CSV row
> csvRows.remove(0);
> map<String,Schema.DisplayType> fieldToFieldTypeMap = new
> map<String,Schema.DisplayType>();
> map<String,String> fieldName = new map<String,String>();
> for (String field : fields) {
> String fieldApiName = field;
> if(fieldMap.get(fieldApiName) == null) {
> fieldApiName = fieldApiName.replaceAll(' ',
> '_').replaceAll('[^a-zA-Z0-9_]', ''); // Convert to alphanumeric and
> underscore
> fieldApiName = fieldApiName.replaceAll('__', '_')+'__c'; // Remove extra
> _s and add __c to end
> }
> fieldToFieldTypeMap.put(field, getFieldType(fieldMap, fieldApiName));
> fieldName.put(field, fieldApiName);
> }
>
> Response output = new Response();
> output.convertedCSVRows = new SObject[]{};
> String tempRepl = '~`~';
> for(String row : csvRows) {
> SObject obj = sObjType.newSObject();
> row = row.replace(curInput.Fsep + '"',curInput.Fsep).replace('"' +
> curInput.Fsep,curInput.Fsep);
> String[] fieldInputValues = row.split(curInput.FSep);
> for (Integer i = 0; i < fields.size(); i++) {
> String fieldValue = fieldInputValues[i].replace(curInput.Fsep +
> '"',curInput.Fsep).replace('"' + curInput.Fsep,curInput.Fsep);
> String Tsep = curInput.Tsep;
> String Dsep = curInput.Dsep;
> String Csymb = curInput.Csymb;
> Schema.DisplayType fieldType = fieldToFieldTypeMap.get(fields[i]);
> try {
> obj.put( fieldName.get(fields[i]), getConvertedFieldValue(fieldType,
> fieldValue, Tsep, Dsep, Csymb) );
> } catch(exception e) {
> obj.put( fieldName.get(fields[i]), null );
> }
> }
> output.convertedCSVRows.add(obj);
> }
> responseList.add(output);
> }
>
>
> }
> return responseList;
> }
>
> static Blob getCsvContent( String contentDocumentId) {
> ContentVersion csvFile = [SELECT
> VersionData
> FROM ContentVersion
> WHERE ContentDocumentId = :contentDocumentId
> AND IsLatest = true
> ];
> System.debug('contentDocumentId is: ' + contentDocumentId);
> System.debug('csvFile is: ' + csvFile);
> return csvFile.VersionData;
> }
>
> /* Converting the CSV file input(BLOB format) to a string */
> static String blobToString(Blob input, String inCharset){
> String hex = EncodingUtil.convertToHex(input);
> System.assertEquals(0, hex.length() & 1);
> final Integer bytesCount = hex.length() >> 1;
> String[] bytes = new String[bytesCount];
> for(Integer i = 0; i < bytesCount; ++i)
> bytes[i] = hex.mid(i << 1, 2);
> return EncodingUtil.urlDecode('%' + String.join(bytes, '%'), inCharset);
> }
>
> static Object getConvertedFieldValue(Schema.DisplayType fieldType, String
> fieldValue, String Tsep, String Dsep, String Csymb){
> Object convertedFieldValue;
>
> switch on fieldType {//Setting Field value based on the field data type
> when BOOLEAN {
> convertedfieldValue = String.isEmpty(fieldValue) ? false :
> Boolean.valueOf(fieldValue);
> }
> when DOUBLE {
> String fValue = fieldValue.replaceAll('^0-9,.','');
> if( fValue.indexOf(Tsep) < 0 && fValue.indexOf(Dsep) < 0 ) {/* both
> WITHOUT Thousands separator and WITHOUT decimal separator */
> /* No conversion needed */
> convertedfieldValue = String.isEmpty(fieldValue) ? null :
> Decimal.valueOf(fValue);
> }
> if( fValue.indexOf(Tsep) < 0 && fValue.indexOf(Dsep) > -1) {/*example:
> change value WITHOUT thousands separator and WITH decimals like 1.234 (US)
> or 1,234 (EU) to 1.234 (US)*/
> /* Convert comma as decimal separator to point */
> convertedfieldValue = String.isEmpty(fieldValue) ? null :
> Decimal.valueOf(fValue.replace(',','.'));
> }
> if( fValue.indexOf(Tsep) > -1 && fValue.indexOf(Dsep) < 0) {/*example:
> change value WITH thousands separator and WITHOUT decimals like 1,234 (US)
> or 1.234 (EU) to 1234 */
> /* Remove comma and point as Thousands separator (no decimals present) */
> convertedfieldValue = String.isEmpty(fieldValue) ? null :
> Decimal.valueOf(fValue.replace(',','').replace('.',''));
> }
> if( fValue.indexOf(Tsep) > -1 && fValue.indexOf(Dsep) > -1 && Tsep == ',')
> {/*example: change value WITH thousand separator and WITH decimals like
> 1,234.56 to 1234.56 (US) */
> /* Remove comma Thousands separator, leave decimal separator unchanged) */
> convertedfieldValue = String.isEmpty(fieldValue) ? null :
> Decimal.valueOf(fValue.replace(',',''));
> }
> if( fValue.indexOf(Tsep) > -1 && fValue.indexOf(Dsep) > -1 && Tsep == '.')
> {/*example: change value WITH thousand separator and WITH decimals like
> 1.234,56 to 1234,56 (EU)*/
> /* Remove point as Thousands separator, change comma to point as decimal
> separator */
> convertedfieldValue = String.isEmpty(fieldValue) ? null :
> Decimal.valueOf(fValue = fValue.replace('.','').replace(',','.'));
> }
> }
> when CURRENCY {
> String fValue =
> fieldValue.replaceAll('^0-9,.','').replace(Csymb,'').replace(' ','');
> if( fValue.indexOf(Tsep) < 0 && fValue.indexOf(Dsep) < 0 ) {/* both
> WITHOUT Thousands separator and WITHOUT decimal separator */
> /* No conversion needed */
> convertedfieldValue = String.isEmpty(fieldValue) ? null :
> Decimal.valueOf(fValue);
> }
> if( fValue.indexOf(Tsep) < 0 && fValue.indexOf(Dsep) > -1) {/*example:
> change value WITHOUT thousands separator and WITH decimals like 1.234 (US)
> or 1,234 (EU) to 1.234 (US)*/
> /* Convert comma as decimal separator to point */
> convertedfieldValue = String.isEmpty(fieldValue) ? null :
> Decimal.valueOf(fValue.replace(',','.'));
> }
> if( fValue.indexOf(Tsep) > -1 && fValue.indexOf(Dsep) < 0) {/*example:
> change value WITH thousands separator and WITHOUT decimals like 1,234 (US)
> or 1.234 (EU) to 1234 */
> /* Remove comma and point as Thousands separator (no decimals present) */
> convertedfieldValue = String.isEmpty(fieldValue) ? null :
> Decimal.valueOf(fValue.replace(',','').replace('.',''));
> }
> if( fValue.indexOf(Tsep) > -1 && fValue.indexOf(Dsep) > -1 && Tsep == ',')
> {/*example: change value WITH thousand separator and WITH decimals like
> 1,234.56 to 1234.56 (US) */
> /* Remove comma Thousands separator, leave decimal separator unchanged) */
> convertedfieldValue = String.isEmpty(fieldValue) ? null :
> Decimal.valueOf(fValue.replace(',',''));
> }
> if( fValue.indexOf(Tsep) > -1 && fValue.indexOf(Dsep) > -1 && Tsep == '.')
> {/*example: change value WITH thousand separator and WITH decimals like
> 1.234,56 to 1234,56 (EU)*/
> /* Remove point as Thousands separator, change comma to point as decimal
> separator */
> convertedfieldValue = String.isEmpty(fieldValue) ? null :
> Decimal.valueOf(fValue = fValue.replace('.','').replace(',','.'));
> }
> }
> when INTEGER {
> /*Remove thousands separator ( , and . like in 1,234 and 1.234 ) */
> convertedfieldValue = String.isEmpty(fieldValue) ? null :
> Integer.valueOf(fieldValue.replaceAll('^0-9',''));
> }
> when DATE {
> /* replace / and : and . in fieldvalue by - like in 2021.12.31 or
> 2021/12/31 or 2021:12:31 or 1.1.2021 or 31/12/2021 or 31:12:2021 */
> String fValue = fieldValue.replaceAll('^0-9','-');
> list<String> lstSplit = fValue.split('-');
> If (fValue.indexOf('-') > 2){
> /* Format found like YYYY-MM-DD */
> Integer Year = Integer.valueOf(lstSplit[0]);
> Integer Month = Integer.valueOf(lstSplit[1]);
> Integer Day = Integer.valueOf(lstSplit[2]);
> convertedfieldValue = String.isEmpty(fieldValue) ? null :
> date.valueOf(Year + '-' + ('0' + Month).Right(2) + '-' + ('0' +
> Day).Right(2));
> }
> else {
> /* Format found like DD-MM-YYYY, convert to YYYY-MM-DD */
> Integer Day = Integer.valueOf(lstSplit[0]);
> Integer Month = Integer.valueOf(lstSplit[1]);
> Integer Year = Integer.valueOf(lstSplit[2]);
> convertedfieldValue = String.isEmpty(fieldValue) ? null :
> date.valueOf(Year + '-' + ('0' + Month).Right(2) + '-' + ('0' +
> Day).Right(2));
> }
> }
> when else {
> convertedfieldValue = fieldValue;
> }
> }
>
> return convertedFieldValue;
> }
>
> @TestVisible
> static Schema.DisplayType getFieldType(map<String, SObjectField> fieldMap,
> String fieldApiName) {
> if(fieldMap.get(fieldApiName) == null) {
> throw new FlowCustomException('Invalid Field Error: '+fieldApiName + '.
> Please double check the API name of the field.');
> }
> return fieldMap.get(fieldApiName).getDescribe().getType();
> }
>
> public class FlowCustomException extends Exception{}
>
> public class Request{
>
> public Request() {
> }
>
> @InvocableVariable(label='Object API Name' required=true)
> public string objectApiName;
>
> @InvocableVariable(label='Separator 1: Fields' description='Choose comma
> (input: , ) or semicolon (input: ; )' required=true)
> public string FSep;
>
> @InvocableVariable(label='Separator 2: Decimals' description='Choose comma
> (input: , ) or semicolon (input: ; )' required=true)
> public string DSep;
>
> @InvocableVariable(label='Separator 3: Thousands' description='Choose
> comma (input: , ) or semicolon (input: ; )' required=true)
> public string TSep;
>
> @InvocableVariable(label='Symbol of Currency' description='Like $, €, £'
> required=true)
> public string Csymb;
>
> @InvocableVariable(label='CSV (single) File ID' required=true)
> public string contentDocumentId;
>
> /* @InvocableVariable(label='CSV List of File IDs')
> public List<String> contentDocumentIdList;
> */ }
>
> public class Response{
> @InvocableVariable(label='List(collection) of Records')
> public List<SObject> convertedCSVRows;
> }
> }
> @isTest
> public with sharing class ConvertAllCSVToRecordsTest {
>
>
> @TestSetup
> static void createData(){
>
> ContentVersion cv = new ContentVersion(
> Title = 'Demo Accounts CSV',
> VersionData = createCSVBlob(),
> PathOnClient = 'DemoAccounts.csv'
> );
> insert cv;
> }
>
> @isTest
> static void testconvertMethodExceptions() {
> ConvertAllCSVToRecords.Request[] flowInputs = createFlowInputsList();
>
> /* 1. FlowInput list size more than 1 exception
> */
> try {
> ConvertAllCSVToRecords.Request input = new
> ConvertAllCSVToRecords.Request();
> flowInputs.add(input);
> ConvertAllCSVToRecords.Response[] flowOutputs =
> ConvertAllCSVToRecords.convert(flowInputs);
> } catch (Exception e) {
> System.assert(e.getMessage().contains('This action is only supported in
> Screen flows.'));
> }
> }
>
> @isTest
> static void testconvertMethod() {
>
> ConvertAllCSVToRecords.Request[] flowInputs = createFlowInputsList();
> Test.startTest();
> ConvertAllCSVToRecords.Response[] flowOutputs =
> ConvertAllCSVToRecords.convert(flowInputs);
> Test.stopTest();
> System.assert(flowOutputs[0].convertedCSVRows.size() == 3);
> }
>
> @isTest
> static void testGetFieldTypeMethodException() {
> try {
> ConvertAllCSVToRecords.getFieldType(Schema.SObjectType.Account.fields.getMap(),'TestField');
>
> } catch (Exception e) {
> System.assert(e.getMessage().contains('Invalid Field Error'));
> }
> }
>
> static Blob createCSVBlob() {
> String csvStringContent;
> Blob csvBlobContent;
> string[] fields = new String[]{
> 'AccountNumber',
> 'AnnualRevenue',
> 'NumberOfEmployees',
> 'Rating',
> 'Name'
> };
>
> String csvHeader = String.join( fields, ',') + '\n';
> csvStringContent = csvHeader;
>
> Account[] accRecords = new Account[]{
> new Account(Name = 'CSV Acc1', AnnualRevenue = NULL, AccountNumber = '',
> NumberOfEmployees = NULL, Rating = 'Hot'),
> new Account(Name = 'CSV Acc2', AnnualRevenue = 1000, AccountNumber =
> '911', NumberOfEmployees = 0, Rating = 'Warm'),
> new Account(Name = 'CSV Acc3', AnnualRevenue = 911007, AccountNumber =
> '007',NumberOfEmployees = 10,Rating = 'Cold')
> };
>
> for(Integer i = 0; i < accRecords.size(); i++) {
> Account a = accRecords[i];
> for(Integer j = 0; j < fields.size(); j++) {
> String field = fields[j];
> csvStringContent = csvStringContent + (a.get(field) == NULL ? '' :
> a.get(field));
> if( j < fields.size()-1) {
> csvStringContent += + ',';
> }
> }
> if( i < accRecords.size()-1) {
> csvStringContent += '\n';
> }
> }
> return Blob.valueOf(csvStringContent);
> }
>
> static ConvertAllCSVToRecords.Request[] createFlowInputsList(){
> String contentDocumentId;
> ContentVersion cv = [Select ContentDocumentId,FileType
> FROM ContentVersion
> Where Title = 'Demo Accounts CSV'
> ];
> ConvertAllCSVToRecords.Request[] flowInputs = new
> ConvertAllCSVToRecords.Request[]{};
> ConvertAllCSVToRecords.Request input = new
> ConvertAllCSVToRecords.Request();
> input.objectApiName = 'Account';
> input.contentDocumentId = cv.ContentDocumentId;
> input.Fsep = ',';
> input.Tsep = ',';
> input.Dsep = '.';
> input.Csymb = '$';
> flowInputs.add(input);
>
> return flowInputs;
> }
>
> }
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/alexed1/LightningFlowComponents/pull/626#issuecomment-792789759>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF4UHX2EEABXAIUENJPOEWLTCTM6PANCNFSM4YV7SJ6A>
> .
>

alexed1 commented:
Replaced by https://github.com/alexed1/LightningFlowComponents/pull/660
